File: lecturer/lib/models/audit.dart
class AuditViolation {
  final String id;
  final String? examSessionId;
  final String? studentSessionId;
  final String studentName;
  final String? deviceId;
  final String violationType;
  final String resourceName;
  final String actionTaken;
  final String? details;
  final DateTime timestamp;

  AuditViolation({
    required this.id,
    this.examSessionId,
    this.studentSessionId,
    required this.studentName,
    this.deviceId,
    required this.violationType,
    required this.resourceName,
    required this.actionTaken,
    this.details,
    required this.timestamp,
  });

  factory AuditViolation.fromJson(Map<String, dynamic> json) {
    return AuditViolation(
      id: json['id'] ?? '',
      examSessionId: json['exam_session_id'],
      studentSessionId: json['student_session_id'],
      studentName: json['student_name'] ?? 'Unknown Student',
      deviceId: json['device_id'],
      violationType: json['violation_type'] ?? 'SECURITY_ALERT',
      resourceName: json['resource_name'] ?? 'Unknown Resource',
      actionTaken: json['action_taken'] ?? 'BLOCKED',
      details: json['details'],
      timestamp: json['timestamp'] != null ? DateTime.parse(json['timestamp']).toUtc() : DateTime.now().toUtc(),
    );
  }
}

File: kasim-reconstructed/backend/app/routers/audit.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from app.database import get_db
from app.models import AuditViolation, ExamSession, User
from app.schemas import AuditViolationCreate, AuditViolationResponse
from app.security import get_current_user


router = APIRouter(prefix="/api/audit", tags=["Audit & Violations"])


@router.post("/violations", response_model=AuditViolationResponse, status_code=status.HTTP_201_CREATED)
def report_violation(
    violation_in: AuditViolationCreate,
    db: Session = Depends(get_db)
):
    """
    Ingests a security violation event dispatched from student desktop agent guards.
    """
    v = AuditViolation(
        exam_session_id=violation_in.exam_session_id,
        student_session_id=violation_in.student_session_id,
        student_name=violation_in.student_name,
        device_id=violation_in.device_id or "Windows Desktop Agent",
        violation_type=violation_in.violation_type,
        resource_name=violation_in.resource_name,
        action_taken=violation_in.action_taken,
        details=violation_in.details
    )
    db.add(v)
    db.commit()
    db.refresh(v)
    return v


@router.get("/violations/{exam_session_id}", response_model=List[AuditViolationResponse])
def get_exam_violations(
    exam_session_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Fetches real-time audit log stream for an exam session.
    """
    if current_user.role != "lecturer":
        raise HTTPException(status_code=403, detail="Only lecturers can access audit logs")

    exam = db.query(ExamSession).filter(ExamSession.id == exam_session_id).first()
    if not exam:
        raise HTTPException(status_code=404, detail="Exam session not found")
    if exam.lecturer_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized to access audit logs for this exam")

    violations = db.query(AuditViolation).filter(
        AuditViolation.exam_session_id == exam_session_id
    ).order_by(AuditViolation.timestamp.desc()).all()

    return violations
